### PR TITLE
Parallel building 2.9

### PR DESCRIPTION
**Pull request by The-going**
#### Description

Add parallel assembly relying on the correct operation of automatic dh commands.

-   export all necessary variables and pass path to src dir
-   fix override_dh_auto_clean
-   PYTHON: Export a variable explicitly to access it
-   dh_auto_build: Provide the possibility of automatic operation

close https://github.com/LinuxCNC/linuxcnc/pull/2579

### CODE REVIEW

File: debian/rules.in [hunk 1]
```diff
@@ -31,40 +33,41 @@ export TIME=$(shell LANG=C date --date='@$(TIMESTAMP)' '+%T')
...
-	dh $@
+	dh $@ -D$(SRCDIR)
 
 override_dh_auto_configure:
 	cd src && ./autogen.sh
-	cd src && PYTHON=/usr/bin/python3 ./configure \
+	dh_auto_configure -- \
 	    --prefix=/usr --sysconfdir=/etc \
+	    --libdir=/usr/lib \
+	    --disable-multiarch \
```
Comment (petterreinholdtsen) [inline debian/rules.in:50]: I believe we should avoid --disable-multiarch.  Any idea how to do that?

Comment (The-going) [reply debian/rules.in:50]: In fact, this is the application of configuration parameters explicitly.
I just brought the problem to the surface.

We can go back to the beginning and write like this:
```make
override_dh_auto_configure:
 	cd src && ./autogen.sh
	cd src && PYTHON=/usr/bin/python3 ./configure \
```
or so:
```make
override_dh_auto_configure:
 	cd src && ./autogen.sh
	dh_configure -D$(SRCDIR) -- \
```
And these configuration parameters will be applied implicitly.

I have not found any description or reference on this issue in the linuxcnc documentation.
There is also no agreement or description about where which files should be located.
So just reading [this: #multiarch](https://www.debian.org/doc/manuals/maint-guide/advanced.en.html#multiarch) and [this: MultiarchSpec](https://wiki.ubuntu.com/MultiarchSpec)

Comment (petterreinholdtsen) [reply debian/rules.in:50]: I guess my problem is that I do not understand how disabling multiarch improve parallelism in the build.  can you explain?

Comment (The-going) [reply debian/rules.in:50]: This does not affect the parallel build in any way.
This affects the parallel execution of the configuration script, but has a small effect.

I will formulate your question in a different way.
What should we additionally do to add support for multiarchitectural assembly?

When we write:
```make
override_dh_auto_configure:
 	cd src && ./autogen.sh
	cd src && PYTHON=/usr/bin/python3 ./configure \
```
This means that the script needs to do literally what is written below and not apply any implicit rules, including parallel assembly and multiarchitecture support.
When we write:
```make
override_dh_auto_configure:
 	cd src && ./autogen.sh
 	dh_auto_configure -- \
 	    --prefix=/usr --sysconfdir=/etc \
	    --libdir=/usr/lib \
	    --disable-multiarch \
```
This means dh_auto_XX to apply all implicit rules except those that are explicitly specified as not to apply.

In other words, before this PR there was no multiarchitectural support
and I just turned it off explicitly in order for the project to be assembled.

**Comment by The-going** [comment]
@petterreinholdtsen What should we additionally do to add support for multiarchitectural assembly?
This is an important issue for me and I will make this support in my repository.
In order for my work to return to the project, the linuxcnc developers need to discuss
and decide whether they will implement this support in the future.
And at least write about these plans in the TODO file.

File: debian/rules.in (hunk 1 above)
Comment (The-going) [reply debian/rules.in]: Going back to that would be a compromise.
```make
override_dh_auto_configure:
 	cd src && ./autogen.sh
	cd src &&  ./configure \
```

**Comment by smoe** [comment]
I just googled a bit for LinuxCNC and multiarch and found this in the context of NVidia and Ethercat, without any deeper understanding, admittedly. Have not seen it discussed on the mailing list over the past years but if technically not completely off with all our kernel modules then this should be supported, indeed. How urgent is this? Could multiarch wait for 3.0? Somehow I expect that the compiler is hard-coded here or there or that some other flags are not passed right at some corner of the source tree.  Since we all have some beef with our Makefiles at some corner, I suggest to elect a volunteer to revision the build process properly.

**Comment by smoe** [comment]
Have just reviewed it all - @petterreinholdtsen still any objections? This would pass me, I actually like it. @The-going, have many thanks! Nice work!

**Comment by The-going** [comment]
> I just googled a bit for LinuxCNC and multiarch

This has to do with the fact that when we have an application and a shared library and the library can be built
for both the i386 or x86_64 architecture and they can both work on the x86_64 architecture.
Similarly with arm, aarch64. Built for these two architectures can run on arm64. In this case,
the location of the shared library files should be in different places:
`/usr/lib/${DEB_HOST_MULTIARCH}/`
That is, replacing this path `/usr/lib/linuxcnc` with this one `/usr/lib/aarch64-linux-gnu/linuxcnc`

Here this question is covered a little: https://wiki.ubuntu.com/MultiarchSpec
I think it's too early to take care of it. I'd rather improve Makefile's a bit first so that the rules for building
a debian package look simple. It has a "mythical" sacred meaning.

**Comment by andypugh** [comment]
Where are we up to with this?

**Comment by smoe** [comment]
@petterreinholdtsen , may we have your OK have this merged?

**Comment by petterreinholdtsen** [comment]
I am fine with this change as it stands now.

**Review by petterreinholdtsen — APPROVED** [review]
